_posts/2016-03-28-Mutual-Problems.md

```diff
@@ -0,0 +1,130 @@
+During [The handshake](https://en.wikipedia.org/wiki/Transport_Layer_Security#TLS_handshake) the following steps are performed:
+
+- A client sends the message which initiates the connection.
+- The server sends its certificate to the client.
+- The client verifies the certificate using the certificate issued by the trusted authority.
+- The server sends the request for the client's certificate.
+- The client sends its certificate to the server.  
+- The server verifies the client's certificate.
+- The server and the client exchange the master secret which is used during the encryption of data.
+- Connection is established.  
+
+Combining my knowledge about the TLS/SSL handshake and my experience from the manual test with `curl` I've assumed that I'm going to need to just three files to implement the client side: **a client's certificate**, **a client's private key** and **a trusted certificate to verify the server's certificate**.
+
+## Java - the problem, the solution and why is it so hard
+
+Because it is quite unusual to use the mutual authentication every day, together with my teammate we asked for the assistance. The first look at the results served by *uncle Google* didn't revealed the complexity behind the implementation, but each click on the results led us to the more and more confusing solutions (some of them where from 90's). To make matters worse we had to use [Apache HttpComponents](https://hc.apache.org/) to implement our connection, but the most of the proposed solutions based on the pure **Java** libraries.  
+
+The knowledge from the internet allows us to establish:
+
+- **Java** cannot use directly any certificates or private keys (like e.g. `curl`)
+- **Java** requires separate files (**Java Keystores**) which can contain original certificates and keys. 
+- We needed a trusted keystore with the certificate required for the server's certificate verification.
+- We needed a keys keystore with the client's certificate and the client's private key.
+
+First we had to create the trusted keystore. We created the keystore with the certificate using the `keytool` command:
+ 
+    $ keytool -import -alias trusted_certificate -keystore trusted.jks -file trusted.crt
+
+We stored in the keystore file `trusted.jks` the certificate `trusted.crt` under the alias `trusted_certificate`. During the execution of this command we were asked to input a password for this keystore. This password we used later to get access to the keystore file.
+
+To create a keystore a few additional steps were required. In most cases you will probably receive two files from the company which issue the client's certificate. The first file will be the client's certificate in the `pem` format. This certificate will be send to the server. The second file is the client's private key (also in the `pem` format) which is used during the handshake to confirm that you are owner of the client's certificate.
```

Comment:
from the company which issue`s`

This certificate will be sen`t`

To confirm that you are `the` owner of...